### PR TITLE
Update frozen actors only when required.

### PR DESCRIPTION
**Pull request by RoosterDragon**
Previously, actors that were visible would refresh their frozen actor state every tick in preparation for the actor becoming hidden, and the frozen actor appearing as a placeholder instead.

By using `ICreatesFrozenActors.OnVisibilityChanged` when can avoid refreshing the state constantly, and instead just refresh it the moment the frozen actor needs to appear. This provides a nice performance improvement on the cost on managing frozen actors.

----

On the RA shellmap, reduces CPU time spent in `UpdateFrozenActor` from 5.1% to <0.1%.

I have also tested that when using the starting a game with Explored Map enabled, that pre-existing frozen structures are visible and targetable, e.g. engineers can be ordered to capture an oil derrick under the fog. We shouldn't have a repeat of the bug fixed in #16194 despite the similar code.

### CODE REVIEW

**Review by Mailaender — DISMISSED** [review]
I can't spot problems in-game.

File: OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs [hunk 1]
```diff
@@ -72,6 +72,18 @@ void INotifyCreated.Created(Actor self)
...
+			// This relies on actor.GetTargetablePositions(), which is also setup up in Created.
+			// Since we can't be sure whether our method will run after theirs, defer by a frame.
+			if (startsRevealed)
+				self.World.AddFrameEndTask(_ =>
+				{
+					for (var playerIndex = 0; playerIndex < frozenStates.Count; playerIndex++)
+						UpdateFrozenActor(frozenStates[playerIndex].FrozenActor, playerIndex);
+				});
+
+			created = true;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs:86]: In the old code this was only set if `startsRevealed` is `true` and was set at the same time the update ran, so should be inside the FrameEndTask as well?

File: OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs [hunk 2]
```diff
@@ -86,9 +98,14 @@ void ICreatesFrozenActors.OnVisibilityChanged(FrozenActor frozen)
 			if (!created)
 				return;
 
-			// Update state visibility to match the frozen actor to ensure consistency within the tick
-			// The rest of the state will be updated by ITick.Tick below
-			frozenStates[frozen.Viewer].IsVisible = !frozen.Visible;
+			// Update state visibility to match the frozen actor to ensure consistency
+			var playerIndex = frozen.Viewer.World.Players.IndexOf(frozen.Viewer);
+			var state = frozenStates[playerIndex];
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs]: Fwiw, `frozenStates` can be indexed directly using `frozen.Viewer` (see the deleted code), so we could defer the `playerIndex` IndexOf-lookup to when it is actually needed (i.e., `isVisible` being true).

File: OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs (hunk 1 above)
Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs:86]: The new logic that's been added to `ICreatesFrozenActors.OnVisibilityChanged` needs to run in both cases. Trying out your suggestion on a map with Explored Map = false results in buildings being invisible. They are invisible both if you have sight and if you do not.

File: OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs (hunk 2 above)
Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs]: Updated

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/db0f75e5e934d81ded485cc09530ab7e37d484bf...7c9fd260841885f03d32db4f3cbecd44ea125e12)

**Pull request merged**